Sources/Core/Math.h

```diff
@@ -27,12 +27,92 @@
 #include <string>
 #include <vector>
 #include <functional>
+#include <mutex>
 
 namespace spades {
 
-	// Make mt_engine and a real dist [0,1] accesible everywhere in the spades namespace
-	extern std::mt19937_64 mt_engine;
-	extern std::uniform_real_distribution<float> real_dist;
+#pragma mark - Random number generation
+	/**
+	 * A thread-safe random number generator satistying
+	 * `UniformRandomBitGenerator`.
+	 */
+	class ThreadSafeRNG {
+	public:
+		using inner_type = std::mt19937_64;
+		using result_type = inner_type::result_type;
+
+		ThreadSafeRNG();
+
+		result_type operator()();
+
+		static constexpr result_type min() { return inner_type::min(); }
+		static constexpr result_type max() { return inner_type::max(); }
+
+	private:
+		inner_type inner;
+		std::mutex mutex;
```

Comment:
Just an idea: what about using an thread-local lazily constructed rng? (Inspired by rust's [rand](https://docs.rs/rand/0.5.5/rand/fn.thread_rng.html) crate).

Reply:
That’s definitely a better idea. (Actually I thought of that at some point but I was about to sleep so I kept going on)